### ISSUE TITLE
Semaphore behavior in Bulkhead strategy (add queueSize and timeout)

### ISSUE THREAD

**Issue by magicprinc** [smallrye/smallrye-fault-tolerance#987]
With JDK21 we have Virtual Threads: they are cheap, and you can block them without any thoughts

Could you add an option to Semaphore based Bulkhead strategies (e.g. timeout: long millis),
which, when enabled, changes the behavior of Semaphore from
`tryAcquire()`
to 
`tryAcquire(long timeout, TimeUnit unit/*MILLISECONDS*/)`

So, if one runs a code with Bulkhead and sets limit to 10, then only 10 threads would run concurrently and other (millions) would be put into "waiting queue"

The application code itself would be much simpler.

Now I have to use plain Semaphore in addition to FT :-(

**Comment by magicprinc** [comment]
Here is an example, 
```
@Test
void _bulkhead () {
  // the other side has set a contractual limit to 2 concurrent connections
  var ft = FaultTolerance.<Integer>create().withDescription("_bulkhead")
    .withBulkhead().limit(2).done()
    .build();

  var total = new AtomicInteger();
  var concurrent = new AtomicInteger();

  // we load tasks (the queue length is controlled) and run them in Virtual Thread
  for (int i=0; i < 20; i++){
    Runnable task = ()->{
      try {
        total.incrementAndGet();
        ft.call(()->{
          concurrent.incrementAndGet();
          try { Thread.sleep(20); } catch (Throwable ignore){}
          System.out.printf("Data sent: total=%s, concurrent=%s%n", total, concurrent);
          concurrent.decrementAndGet();
          return null;
        });
      } catch (Exception e) {
        System.out.printf("FT.call failed: total=%s%n", total);
      }
    };

    JThread.VT.execute(task);// run in VT using VT executor
  }
}
```

here is the result
```
FT.call failed: total=16
FT.call failed: total=18
FT.call failed: total=19
FT.call failed: total=20
FT.call failed: total=20
FT.call failed: total=20
FT.call failed: total=20
FT.call failed: total=20
FT.call failed: total=20
FT.call failed: total=20
FT.call failed: total=20
FT.call failed: total=20
FT.call failed: total=20
FT.call failed: total=20
FT.call failed: total=20
FT.call failed: total=20
FT.call failed: total=20
FT.call failed: total=20
Data sent: total=20, concurrent=2
Data sent: total=20, concurrent=1
```

**Comment by magicprinc** [comment]
We live in VT world now: you usually don't need async and CompletableFuture 🥳

**Comment by magicprinc** [comment]
~ "Bulkhead with timeout"

```
@Test @SneakyThrows
void _bulkhead () {
  // the other side has set a contractual limit to 2 concurrent connections
  var ft = new Semaphore(2);

  var total = new AtomicInteger();
  var concurrent = new AtomicInteger();
  var endTest = new CountDownLatch(20);

  // we load tasks (the queue length is controlled) and run them in Virtual Thread
  for (int i=0; i < 20; i++){
    Runnable task = ()->{
      try {
        total.incrementAndGet();
        ft.tryAcquire(20, TimeUnit.SECONDS);// dirty quick demo: result ignored
        try {
          concurrent.incrementAndGet();
          Thread.sleep(20);
          System.out.printf("Data sent: total=%s, concurrent=%s%n", total, concurrent);
          concurrent.decrementAndGet();
        } finally {
          ft.release();
        }
      } catch (Exception e) {
        System.out.printf("FT.call failed: total=%s%n", total);
      } finally {
        endTest.countDown();
      }
    };

    JThread.VT.execute(task);// run in VT using VT executor
  }
  endTest.await();
}
```

result
```
Data sent: total=20, concurrent=2
Data sent: total=20, concurrent=1
Data sent: total=20, concurrent=2
Data sent: total=20, concurrent=2
Data sent: total=20, concurrent=2
Data sent: total=20, concurrent=2
Data sent: total=20, concurrent=2
Data sent: total=20, concurrent=2
Data sent: total=20, concurrent=2
Data sent: total=20, concurrent=2
Data sent: total=20, concurrent=2
Data sent: total=20, concurrent=2
Data sent: total=20, concurrent=2
Data sent: total=20, concurrent=2
Data sent: total=20, concurrent=2
Data sent: total=20, concurrent=2
Data sent: total=20, concurrent=2
Data sent: total=20, concurrent=2
Data sent: total=20, concurrent=2
Data sent: total=20, concurrent=1
```

**Comment by Ladicek** [comment]
Don't let anyone tell you that virtual threads are some form of a panacea; they are still just threads. The only thing that might possibly change in the future is that we will no longer need to deal with non-blocking concurrency (aka event loops or "reactive"); blocking concurrency (aka multi-threading) will be enough. But we're certainly not in that world yet; it will take years for the ecosystem to adapt.

Some people equate the word "asynchronous" with one or the other form of concurrency I mentioned above. In SmallRye Fault Tolerance, I use the word "asynchronous" in a more abstract manner, close to the dictionary definition of "not happening at the same time". 

If an operation is asynchronous, it means its result is not available right here right now; the idea of reifying that concept into a _future_ is decades old. As it turns out, if we restrict ourselves to the JDK, the best type for representing futures is currently `CompletionStage` / `CompletableFuture`, because it supports both styles of concurrency. So that's what SmallRye Fault Tolerance uses. Maybe at some point, this won't be necessary, but as I mentioned above, we're years away from that.

Now, here's a piece of code that uses bulkhead with a queue (which requires the operation to be asynchronous, otherwise there's no way to implement queueing):

```java
@Test
public void test() throws Exception {
    var ft = FaultTolerance.<Integer>createAsync()
            .withDescription("_bulkhead")
            .withBulkhead().limit(2).queueSize(100).done()
            .build();

    var total = new AtomicInteger();
    var concurrent = new AtomicInteger();
    var endTest = new CountDownLatch(20);

    for (int i = 0; i < 20; i++) {
        Runnable task = () -> {
            try {
                total.incrementAndGet();
                ft.call(() -> {
                    concurrent.incrementAndGet();
                    try {
                        Thread.sleep(20);
                    } catch (Throwable ignore) {
                    }
                    System.out.printf("Data sent: total=%s, concurrent=%s%n", total, concurrent);
                    concurrent.decrementAndGet();
                    return null;
                });
            } catch (Exception e) {
                System.out.printf("FT.call failed: total=%s%n", total);
            } finally {
                endTest.countDown();
            }
        };

        executor.execute(task);
    }

    endTest.await();
}
```

Apparently, returning `null` works in this particular case, but I wouldn't rely on it for anything real. I really would return `CompletableFuture.completedStage(null);`.

If I rely on SmallRye Fault Tolerance to submit the tasks to an executor on its own, instead of doing it manually, that code becomes:

```java
@Test
public void test() throws Exception {
    var ft = FaultTolerance.<Integer>createAsync()
            .withDescription("_bulkhead")
            .withBulkhead().limit(2).queueSize(100).done()
            .withThreadOffload(true)
            .build();

    var total = new AtomicInteger();
    var concurrent = new AtomicInteger();
    var endTest = new CountDownLatch(20);

    for (int i = 0; i < 20; i++) {
        ft.call(() -> {
            total.incrementAndGet();
            concurrent.incrementAndGet();
            try {
                Thread.sleep(20);
            } catch (Throwable ignore) {
            }
            System.out.printf("Data sent: total=%s, concurrent=%s%n", total, concurrent);
            concurrent.decrementAndGet();
            endTest.countDown();
            return CompletableFuture.completedStage(null);
        });
    }

    endTest.await();
}
```

This works perfectly well.

**Comment by Ladicek** [comment]
Ah OK, so you want the queued tasks to have a timeout? And if the task is in the queue for too long, it should fail immediately? That seems like a reasonable request.

**Comment by magicprinc** [comment]
VT are "async". But they are sync somewhere in the depth.
They are good enough for me.
Oracle's Helidon Nima has completely removed Netty and uses plain old sockets + Virtual Threads.

> otherwise there's no way to implement queueing
There is:
limit + queue ≈ tryAcquire with timeout

We started to remove "manual async" code and usually in reduces code lines 2-4 times

**Comment by Ladicek** [comment]
Ah I see. If you have 20 threads competing for a semaphore that only hands out 2 permits, you indeed have an implicit form of queueing, but at the cost of having 20 threads. That may be fine with virtual threads, but it definitely is not fine with platform threads.

Now, when it comes to

> limit + queue ≈ tryAcquire with timeout

that's not true. You don't need the timeout there, that doesn't really change anything when it comes to the queueing semantics. It is an extra cherry on top.

In SmallRye Fault Tolerance, bulkheads implement an _explicit_ form of queueing. Tasks are put into an actual queue and are not submitted to an executor until they obtain a permit. (Aside: in the annotation-based API, this is only true for `CompletionStage`-returning tasks and not for `Future`-returning tasks, but the programmatic API doesn't support `Future`-returning tasks at all.)

**Comment by magicprinc** [comment]
> You don't need the timeout there, that doesn't really change anything when it comes to the queueing semantics

Without timeout, FT Bulkhead simply fails the task.
See results in example 1 (current implementation) and in example 2 ("Bulkhead with timeout")

You can definitely see a "queue" in example 2

> In SmallRye Fault Tolerance, bulkheads implement an explicit form of queueing

New times, new rules maybe?

With VT you can delegate queuing to them. It is an official position of JDK architects.

**Comment by Ladicek** [comment]
Actually one more thing, I probably didn't express that clearly above. The usage of `CompletionStage` in SmallRye Fault Tolerance doesn't mean you're _required_ to use non-blocking concurrency. If you do thread offloads (either manually, or ask SmallRye Fault Tolerance to do it for you), you can use blocking concurrency just fine. In that case, you will only create already complete `CompletionStage`s (`CompletableFuture.completedStage()` and `failedStage()`). If you want to block on the `CompletionStage` returned by SmallRye Fault Tolerance, you can call `toCompletableFuture()`, and that's guaranteed to work. This indeed entails certain amount of code overhead, but not very big I believe.

**Comment by Ladicek** [comment]
> Without timeout, FT Bulkhead simply fails the task.

Ah ah ah, you're right, my bad. I didn't look properly at how the `Semaphore` methods differ.

> New times, new rules maybe?
> With VT you can delegate queuing to them. It is an official position of JDK architects.

SmallRye Fault Tolerance has to work in the actual real world, which will take years to fully adopt virtual threads. Sorry, but this is just the sad state of reality.

**Comment by magicprinc** [comment]
Now I see your point.
It should definitely help!

Could you nevertheless add timeout to Bulkhead to 
1) limit time in the queue
2) prepare for great code simplification in the future (complete removal of Async part) 
?

**Comment by Ladicek** [comment]
Looking at `Semaphore` again, you actually _don't_ need timeout to achieve queueing. You just need `Semaphore.acquire()` instead of `tryAcquire()`. The timeout is useful, but not strictly necessary when it comes to achieving the queueing semantics.

Now, supporting timeout for the time spent in queue is a good idea, but doesn't feel like a super high priority right now. It would be relatively complex to implement in the bulkheads that have an explicit queue, actually.

**Comment by magicprinc** [comment]
`Semaphore.acquire()` doesn't have timeout and I am afraid of this time to be too long...
I don't usually like blocking operations without timeout

**Comment by Ladicek** [comment]
That is indeed reasonable.

**Comment by magicprinc** [comment]
Ok, if we call it time-in-the-queue timeout, it makes things complicated.

It is timeout to acquire workSemaphore. 
It has sense even in case of CF version.
If you have spikes in the traffic, then it is better to have some timeout, don't fail immediately.

It helps to smooth out the peaks, keep order of messages, etc

**Comment by magicprinc** [comment]
Great job! 🥳

Is it still possible to add acquireTimeout to SemaphoreBulkhead and FutureThreadPoolBulkhead (is it still used?)? 🙏😅

then
```java
@Test
public void test() throws Exception {
    var ft = FaultTolerance.<Integer>create()
            .withDescription("_bulkhead")
            .withBulkhead().limit(2).acquireTimeout(5, MINUTES)/* currently fails in case of CompletionStageThreadPoolBulkhead */.done()
            .build();

    var total = new AtomicInteger();
    var concurrent = new AtomicInteger();

    for (int i = 0; i < 20; i++) {
        ft.run(() -> {
            total.incrementAndGet();
            concurrent.incrementAndGet();
            try {
                Thread.sleep(20);
            } catch (Throwable ignore) {
            }
            System.out.printf("Data sent: total=%s, concurrent=%s%n", total, concurrent);
            concurrent.decrementAndGet();
        });
    }
}
```

**Comment by Ladicek** [comment]
`FutureThreadPoolBulkhead` is not used in the programmatic API, only in the annotation-based API.

I'm not going to add a timeout to bulkheads at the moment. It feels like a useful feature, but it also seems there's a big potential for scope creep. I'm going to recommend that if you want concurrency limiter with a queue, you should treat the guarded actions as asynchronous.

**Comment by magicprinc** [comment]
Forget about queue for a moment, let's say: I don't want sync Bulkhead to fail immediately: I am ready to wait

I am ready to wait with retry (and I do: there is Thread.sleep there) and am ready to wait with Bulkhead

**Comment by Ladicek** [comment]
That's exactly the scope creep I want to avoid. Bulkhead (aka concurrency limiter) already has a concept of a queue, the concept of a timeout somewhat conflicts with that. It actually feels like allowing `queueSize` tasks to sleep and start rejecting only after that is what you need (and that's actually what `FutureThreadPoolBulkhead` already implements, using 2 semaphores). There might be a way to allow that.

**Comment by magicprinc** [comment]
Idea with 2 semaphores in sync blocking code has sense!
I feel it can be implemented and do exactly what VT users need.

**Comment by magicprinc** [comment]
FutureThreadPoolBulkhead works as expected (except of course the return type :-)

```
@Test @SneakyThrows
void bulkheadWithQueue () {
  var total = new AtomicInteger();
  var concurrent = new AtomicInteger();

  // the other side has set a contractual limit to 2 concurrent connections
  var ft = new FutureThreadPoolBulkhead<Integer>(InvocationContext::call, "bhq", 2, 20){
    @Override @SneakyThrows
    public Future<Integer> apply (InvocationContext<Future<Integer>> ctx) {
      total.incrementAndGet();
      return super.apply(ctx);
    }
  };

  var futures = ConcurrentHashMap.<Future<Integer>>newKeySet();

  // we load tasks (the queue length is controlled) and run them in Virtual Thread
  for (int i=0; i < 20; i++){
    int index = i;
    Callable<Future<Integer>> task = ()->{
      try {
        concurrent.incrementAndGet();
        try { Thread.sleep(20); } catch (Throwable ignore){}
        System.out.printf("Data sent: total=%s, concurrent=%s%n", total, concurrent);
        concurrent.decrementAndGet();
      } catch (Exception e) {
        System.out.printf("FT.call failed: total=%s%n", total);
      }
      return Futures.immediateFuture(index);
    };
    var ctx = new InvocationContext<Future<Integer>>(task);

    JThread.VT.execute(()->{
      var f = ft.apply(ctx);
      futures.add(f);
    });// run in VT using VT executor
  }

  while (futures.size() < 20) {
    Thread.sleep(10);
  }

  for (var f : futures){
    System.out.println(f.get());
  }
}
```

result

```
Data sent: total=20, concurrent=2
Data sent: total=20, concurrent=1
Data sent: total=20, concurrent=2
Data sent: total=20, concurrent=2
Data sent: total=20, concurrent=2
Data sent: total=20, concurrent=2
Data sent: total=20, concurrent=2
Data sent: total=20, concurrent=2
Data sent: total=20, concurrent=2
Data sent: total=20, concurrent=2
Data sent: total=20, concurrent=2
Data sent: total=20, concurrent=2
Data sent: total=20, concurrent=2
Data sent: total=20, concurrent=2
Data sent: total=20, concurrent=2
Data sent: total=20, concurrent=2
Data sent: total=20, concurrent=2
Data sent: total=20, concurrent=2
Data sent: total=20, concurrent=2
Data sent: total=20, concurrent=1
2
9
15
13
1
...
```

**Comment by magicprinc** [comment]
I have made a PR for active JDK 21+ VT users.

Sync/Blocking Bulkhead with Queue and Limit.

Could you please review/rewrite and accept this PR.

This is the right Bulkhead behavior and I/We/Everybody really need it 🙏🤲🤝

https://github.com/smallrye/smallrye-fault-tolerance/pull/1015

**Comment by magicprinc** [comment]
@Ladicek Please! 🙏😥
As you see no timeouts 🤝 (I have managed to emulate them)
But I need two semaphores very badly…
https://github.com/smallrye/smallrye-fault-tolerance/pull/1015 → this is your own `io.smallrye.faulttolerance.core.bulkhead.FutureThreadPoolBulkhead`, but without Future and included in smallrye-fault-tolerance-standalone

**Comment by magicprinc** [comment]
Now, PR has unit tests and example use case: how FT can be used in VT environment.

**Comment by magicprinc** [comment]
🙏

**Comment by magicprinc** [comment]
Concurrent libraries have such functionality
https://github.com/smallrye/smallrye-fault-tolerance/issues/987

I have used your class as foundation and adopted for JDK 21 VT reality.
Changes are minimal and easy to understand.

Unit Tests are included.
https://github.com/smallrye/smallrye-fault-tolerance/pull/1015/files

**Comment by magicprinc** [comment]
@Ladicek Please! 🙏🙏🙏😭

**Comment by Ladicek** [comment]
Hi @magicprinc, sorry for keeping this open for so long and not doing anything about it. It's been a low priority issue for a long time.

I submitted #1062, which is more explicit about what's going on, because the new bulkhead strategy cannot be used outside of virtual threads. So you have to call `enableVirtualThreadsQueueing()`. Hope that's reasonable for you.
